cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -0,0 +1,335 @@
+import org.apache.tephra.TransactionFailureException;
+import org.apache.tephra.TransactionSystemClient;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Polls the JobQueue, checks the jobs for constraint satisfaction, and launches them.
+ */
+class ConstraintCheckerService extends AbstractExecutionThreadService {
+  private static final Logger LOG = LoggerFactory.getLogger(ConstraintCheckerService.class);
+
+  private final List<ConstraintCheckerThread> constraintCheckerThreads;
+  private final Transactional transactional;
+  private final DatasetFramework datasetFramework;
+  private final Store store;
+  private final ProgramLifecycleService lifecycleService;
+  private final PropertiesResolver propertiesResolver;
+  private final NamespaceQueryAdmin namespaceQueryAdmin;
+  private final CConfiguration cConf;
+  private ScheduleTaskRunner taskRunner;
+  private ListeningExecutorService taskExecutorService;
+
+  @Inject
+  ConstraintCheckerService(Store store,
+                           ProgramLifecycleService lifecycleService, PropertiesResolver propertiesResolver,
+                           NamespaceQueryAdmin namespaceQueryAdmin,
+                           CConfiguration cConf,
+                           DatasetFramework datasetFramework,
+                           TransactionSystemClient txClient) {
+    this.store = store;
+    this.lifecycleService = lifecycleService;
+    this.propertiesResolver = propertiesResolver;
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
+    this.cConf = cConf;
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(new MultiThreadDatasetCache(
+        new SystemDatasetInstantiator(datasetFramework), txClient,
+        NamespaceId.SYSTEM, ImmutableMap.<String, String>of(), null, null)),
+      RetryStrategies.retryOnConflict(20, 100)
+    );
+    this.datasetFramework = datasetFramework;
+    this.constraintCheckerThreads = new ArrayList<>();
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    taskExecutorService = MoreExecutors.listeningDecorator(
+      Executors.newCachedThreadPool(Threads.createDaemonThreadFactory("constraint-checker-task")));
+    taskRunner = new ScheduleTaskRunner(store, lifecycleService, propertiesResolver,
+                                        taskExecutorService, namespaceQueryAdmin, cConf);
+  }
+
+  @Override
+  protected void run() {
+    LOG.info("Start running ConstraintCheckerService");
+    if (!isRunning()) {
+      return;
+    }
+
+    for (int partition = 0; partition < getNumPartitions(); partition++) {
+      constraintCheckerThreads.add(new ConstraintCheckerThread(partition));
```

Comment:
is there a more elegant way to do this? using an executor... perhaps shutdown would be easier?